lib/xgu/xgu.h

```diff
@@ -0,0 +1,148 @@
+
+typedef enum {
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_D3D     0
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S1         1
+  XGU_FLOAT = NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_F,
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_UB_OGL     4
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_S32K       5
+  //FIXME: define NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE_CMP        6
+} XguVertexArrayType;
+
+typedef enum {
+  XGU_END = NV097_SET_BEGIN_END_OP_END, // FIXME: Disallow this one?
+  XGU_POINTS = NV097_SET_BEGIN_END_OP_POINTS,
+  XGU_LINES = NV097_SET_BEGIN_END_OP_LINES,
+  XGU_LINE_LOOP = NV097_SET_BEGIN_END_OP_LINE_LOOP,
+  XGU_LINE_STRIP = NV097_SET_BEGIN_END_OP_LINE_STRIP,
+  XGU_TRIANGLES = NV097_SET_BEGIN_END_OP_TRIANGLES,
+  XGU_TRIANGLE_STRIP = NV097_SET_BEGIN_END_OP_TRIANGLE_STRIP,
+  XGU_TRIANGLE_FAN = NV097_SET_BEGIN_END_OP_TRIANGLE_FAN,
+  XGU_QUADS = NV097_SET_BEGIN_END_OP_QUADS,
+  XGU_QUAD_STRIP = NV097_SET_BEGIN_END_OP_QUAD_STRIP,
+  XGU_POLYGON = NV097_SET_BEGIN_END_OP_POLYGON
+} XguPrimitiveType;
+
+typedef enum {
+  XGU_VERTEX_ARRAY = NV2A_VERTEX_ATTR_POSITION,
+  XGU_NORMAL_ARRAY = NV2A_VERTEX_ATTR_NORMAL,
+  XGU_COLOR_ARRAY = NV2A_VERTEX_ATTR_DIFFUSE,
+  XGU_SECONDARY_COLOR_ARRAY = NV2A_VERTEX_ATTR_SPECULAR,
+  XGU_FOG_ARRAY = NV2A_VERTEX_ATTR_FOG,
+  XGU_TEXCOORD0_ARRAY = NV2A_VERTEX_ATTR_TEXTURE0,
+  XGU_TEXCOORD1_ARRAY = NV2A_VERTEX_ATTR_TEXTURE1,
+  XGU_TEXCOORD2_ARRAY = NV2A_VERTEX_ATTR_TEXTURE2,
+  XGU_TEXCOORD3_ARRAY = NV2A_VERTEX_ATTR_TEXTURE3
+} XguVertexArray;
+
+static inline
+DWORD* push_command(DWORD* p, DWORD command, int nparams) {
+    pb_push(p++, command, nparams);
+    return p;
+}
+
+static inline
+DWORD* push_parameter(DWORD* p, DWORD parameter) {
+    *p++ = parameter;
+    return p;
+}
+
+static inline
+DWORD* push_boolean(DWORD* p, bool enabled) {
+  return push_parameter(p, enabled ? 1 : 0);
+}
+
+static inline
+DWORD* push_command_boolean(DWORD* p, DWORD command, bool enabled) {
+  p = push_command(p, command, 1);
+  p = push_boolean(p, enabled);
+  return p;
+}
+
+static inline
+DWORD* push_command_parameter(DWORD* p, DWORD command, DWORD parameter) {
+  p = push_command(p, command, 1);
+  p = push_parameter(p, parameter);
+  return p;
+}
+
+static inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+  // Force people to use xgu_end instead
+  assert(type != NV097_SET_BEGIN_END_OP_END);
+
+  return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+static inline
+DWORD* xgu_end(DWORD* p) {
+  return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_clear_color(DWORD* p, DWORD color) {
+    p = push_command_parameter(p, NV097_SET_COLOR_CLEAR_VALUE, color);
+    return p;
+}
+
+inline
+DWORD* xgu_clear(DWORD* p) {
+    /* Should probably do some magic related to 
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_HORIZ and
+     * NV20_TCL_PRIMITIVE_3D_CLEAR_VALUE_DEPTH
```

Comment:
Yes, sort of; let's look at the GPU command and its parameters:

```c
#   define NV097_CLEAR_SURFACE                                0x00001D94
#       define NV097_CLEAR_SURFACE_Z                              (1 << 0)
#       define NV097_CLEAR_SURFACE_STENCIL                        (1 << 1)
#       define NV097_CLEAR_SURFACE_COLOR                          0x000000F0
#       define NV097_CLEAR_SURFACE_R                                (1 << 4)
#       define NV097_CLEAR_SURFACE_G                                (1 << 5)
#       define NV097_CLEAR_SURFACE_B                                (1 << 6)
#       define NV097_CLEAR_SURFACE_A                                (1 << 7)
```

So you have the command `NV097_CLEAR_SURFACE` (`xgu_clear_surface`) and it has a couple of booleans / masks to control which color channels are erased:
`xgu_clear_surface(bool z, bool stencil, bool r, bool g, bool b, bool a)`

In the GL world, this is similar, where `glClear` takes a bitmask combination of:
```c
GL_COLOR_BUFFER_BIT
GL_DEPTH_BUFFER_BIT
GL_ACCUM_BUFFER_BIT
GL_STENCIL_BUFFER_BIT
```
so you could also rewrite the GL function as:

```c
void glClear_wrapper(bool color, bool depth, bool accum, bool stencil) {
  int flags = 0;
  if (color) { flags |= GL_COLOR_BUFFER_BIT; } // Equivalent to NV097_CLEAR_SURFACE_{R,G,B,A} combined (aka NV097_CLEAR_SURFACE_COLOR)
  if (depth) { flags |= GL_DEPTH_BUFFER_BIT; } // Equivalent to NV097_CLEAR_SURFACE_Z
  if (accum) { flags |= GL_ACCUM_BUFFER_BIT; } // GL software garbage that's not supported by hardware directly
  if (stencil) { flags |= GL_STENCIL_BUFFER_BIT; } // Equivalent to NV097_CLEAR_SURFACE_STENCIL
  glClear(flags);
}

Wether you pick an API like `glClear` (bitmasks) or `glClear_wrapper` (booleans) is a style choice. However, using the bitmasks makes for more readable code:

`glClear_wrapper(true, true, false, false)` vs `glClear(GL_COLOR_BUFFER_BIT | GL_STENCIL_BUFFER_BIT)`.

in the XGU world it would likely be:

`xgu_clear_surface(true, false, true, true, true, true)` vs `xgu_clear_surface(XGU_CLEAR_COLOR | XGU_CLEAR_DEPTH)`

I'd recommend using the bitmask approach ;)